File: backend/app/domain/schemas/context_builds.py
from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, Field


class ContextBuildRequest(BaseModel):
    conversation_id: str | None = Field(default=None, max_length=64)
    task_id: str | None = Field(default=None, max_length=64)
    query: str = Field(min_length=1, max_length=50_000)
    token_budget: int = Field(default=16_000, ge=256, le=500_000)
    agent_id: str = Field(default="main_agent", max_length=80)
    provider_id: str = Field(default="", max_length=80)
    model_id: str = Field(default="", max_length=200)
    allowed_sensitivity: list[
        Literal["public", "normal", "private", "sensitive", "restricted"]
    ] = Field(default_factory=lambda: ["public", "normal", "private"])
    debug_manifest: bool = False


class ContextEvidenceView(BaseModel):
    kind: str
    target_id: str
    title: str
    content: str
    source_event_id: str
    scope: str
    confidence: float
    status: str
    retrieval_reason: str
    trust: str
    score: float
    component_scores: dict[str, float] = Field(default_factory=dict)


class ContextBuildView(BaseModel):
    context_build_id: str
    trace_id: str | None = None
    task_state: dict[str, Any] | None = None
    recent_context: list[dict[str, Any]] = Field(default_factory=list)
    memories: list[ContextEvidenceView] = Field(default_factory=list)
    episodes: list[dict[str, Any]] = Field(default_factory=list)
    project_decisions: list[dict[str, Any]] = Field(default_factory=list)
    file_chunks: list[dict[str, Any]] = Field(default_factory=list)
    learning_states: list[dict[str, Any]] = Field(default_factory=list)
    strategies: list[dict[str, Any]] = Field(default_factory=list)
    tool_candidates: list[dict[str, Any]] = Field(default_factory=list)
    provider_messages: list[dict[str, Any]] = Field(default_factory=list)
    context_manifest: list[dict[str, Any]] = Field(default_factory=list)
    section_tokens: dict[str, int] = Field(default_factory=dict)
    total_tokens: int = 0
    package_hash: str
    excluded: dict[str, int] = Field(default_factory=dict)
    degraded_modes: list[str] = Field(default_factory=list)

File: frontend/src/types/context-builds.ts
export interface ContextEvidence {
  kind: string
  target_id: string
  title: string
  content: string
  content_hash: string
  source_event_id: string
  scope: string
  confidence: number
  status: string
  retrieval_reason: string
  trust: string
  score: number
  component_scores: Record<string, number>
  reason_codes: string[]
  token_cost: number
  manifest_status: string
}

export interface ContextBuildRequest {
  conversation_id?: string
  message_id?: string
  task_id?: string
  query: string
  token_budget?: number
  agent_id?: string
  provider_id?: string
  model_id?: string
  allowed_sensitivity?: string[]
  debug_manifest?: boolean
}

export interface ContextBuildView {
  context_build_id: string
  task_state: Record<string, unknown> | null
  recent_context: Array<Record<string, unknown>>
  memories: ContextEvidence[]
  candidate_memories: ContextEvidence[]
  retrieved_memories: ContextEvidence[]
  selected_memories: ContextEvidence[]
  injected_memories: ContextEvidence[]
  excluded_memories: ContextEvidence[]
  truncated_memories: ContextEvidence[]
  candidate_count: number
  retrieved_count: number
  selected_count: number
  injected_count: number
  excluded_count: number
  truncated_count: number
  manifest_status: string
  episodes: Array<Record<string, unknown>>
  project_decisions: Array<Record<string, unknown>>
  file_chunks: Array<Record<string, unknown>>
  learning_states: Array<Record<string, unknown>>
  strategies: Array<Record<string, unknown>>
  tool_candidates: Array<Record<string, unknown>>
  provider_messages: Array<Record<string, unknown>>
  context_manifest: Array<Record<string, unknown>>
  section_tokens: Record<string, number>
  total_tokens: number
  package_hash: string
  excluded: Record<string, number>
  degraded_modes: string[]
}

export interface ContextManifestReceipt {
  context_build_id: string
  session_id: string | null
  message_id: string | null
  status: string
  candidate_ids: string[]
  retrieved_ids: string[]
  selected_ids: string[]
  injected_ids: string[]
  excluded_ids: string[]
  truncated_ids: string[]
  reason_codes: Record<string, string[]>
  excluded_counts: Record<string, number>
  section_tokens: Record<string, number>
  total_tokens: number
  injected_tokens: number
  package_hash: string
  created_at: string
}
